### PR TITLE
Permit loose manifest priority

### PR DESCRIPTION
**Pull request by Forgind**
### Description
If the user opts into using workload sets, installs a workload set, then tries to switch back to using loose manifests, we correctly permit installing or updating to loose manifests, but we do not garbage collect the workload sets, both because garbage collection isn't yet implemented for workload sets and because we would keep the latest workload set regardless. That workload set is still findable, therefore, and when the user tries `dotnet workload --version`, `dotnet workload update --print-rollback`, or anything else that uses the resolver to find workloads, it uses the workload set instead of correctly switching back to loose manifests.

### Customer Impact
Customers who ever use workload sets can never use loose manifests again unless they follow manual steps to delete their workload set(s).

### Regression
No; the resolver always worked this way, but workload sets didn't exist, so it wasn't an issue.

### Risk
Low. Relatively few people are using workload sets, and this is changing behavior to align with what the user explicitly requested (or got by using a rollback file), so it changes behavior to align with user expectations. If they'd switched to loose manifests by accident (somehow?), they can always switch back very easily.

### Testing
Changed automated tests to account for the change and walked through one in a debugger to make sure it did what I'd expect.

### CODE REVIEW

**Review by dsplaisted — COMMENTED** [review]
Initially I didn't fully understand your change, so I made a version of the change myself: https://github.com/dsplaisted/sdk/commit/3c5476936a335d37851773a100076ea9fded3de1

On looking at your change a second time, I think it generally works (I had missed the early return statement).  However it looks like your change may not respect global.json or a workload set version passed in via the constructor, both of which it should do.

Since this is a change to the resolver which goes into full VS, I think we should consider taking this change for 8.0.4xx if we can. @marcpopMSFT

File: test/Microsoft.NET.Sdk.WorkloadManifestReader.Tests/SdkDirectoryWorkloadManifestProviderTests.cs [hunk 1]
```diff
@@ -1254,6 +1255,12 @@ private void CreateMockManifest(string manifestRoot, string featureBand, string
 
         private void CreateMockWorkloadSet(string manifestRoot, string featureBand, string workloadSetVersion, string workloadSetContents)
         {
+            // To prepare the resolver to work with workload sets, we need to specify 'workload sets' in the install state file.
+            var installStateLocation = WorkloadInstallType.GetInstallStateFolder(new SdkFeatureBand(featureBand), Path.GetDirectoryName(manifestRoot)!);
+            var installState = InstallStateContents.FromPath(installStateLocation);
+            installState.UseWorkloadSets = true;
+            File.WriteAllText(installStateLocation, installState.ToString());
```
Comment (Forgind) [inline test/Microsoft.NET.Sdk.WorkloadManifestReader.Tests/SdkDirectoryWorkloadManifestProviderTests.cs:1262]:
```suggestion
            var installStateLocation = WorkloadInstallType.GetInstallStateFolder(new SdkFeatureBand(featureBand), Path.GetDirectoryName(manifestRoot)!);
            var installStateFilePath = Path.Combine(installStateLocation, "default.json");
            var installState = InstallStateContents.FromPath(installStateFilePath);
            installState.UseWorkloadSets = true;
            Directory.CreateDirectory(installStateLocation);
            File.WriteAllText(installStateFilePath, installState.ToString());
```

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs [hunk 2]
```diff
@@ -147,28 +147,24 @@ public void RefreshWorkloadManifests()
                 }
             }
 
+            _installStateFilePath = Path.Combine(WorkloadInstallType.GetInstallStateFolder(_sdkVersionBand, _sdkRootPath), "default.json");
+            var installState = InstallStateContents.FromPath(_installStateFilePath);
```
Comment (marcpopMSFT) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:151]: what happens if the file doesn't exist or is empty/malformed?

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs [hunk 3]
```diff
@@ -147,28 +147,24 @@ public void RefreshWorkloadManifests()
...
                     }
-
-                    //  Note: It is possible here to have both a workload set and loose manifests listed in the install state.  This might happen if there is a
-                    //  third-party workload manifest installed that's not part of the workload set
-                    _manifestsFromInstallState = installState.Manifests is null ? null : WorkloadSet.FromDictionaryForJson(installState.Manifests!, _sdkVersionBand);
-                    _installStateFilePath = installStateFilePath;
                 }
+
+                //  Note: It is possible here to have both a workload set and loose manifests listed in the install state.  This might happen if there is a
+                //  third-party workload manifest installed that's not part of the workload set
```
Comment (marcpopMSFT) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:163]: good catch. This will be true of all Tizen customers

Comment (marcpopMSFT) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:163]: ahh, that was from the prior version of the file. needed to hide whitespace from GH

**Comment by marcpopMSFT** [comment]
So the primary change is adding installState.UseWorkloadSets == true to the check, correct? That seems pretty safe and isolated to add. Any concerns with taking this for 8.0.4xx?

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs (hunk 2 above)
Comment (Forgind) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:151]: InstallStateContents has a File.Exists check. It being malformed isn't supported, especially as users are not expected to modify their install states directly, so that's an error.

**Comment by Forgind** [comment]
> So the primary change is adding installState.UseWorkloadSets == true to the check, correct? That seems pretty safe and isolated to add. Any concerns with taking this for 8.0.4xx?

Not from me. @dsplaisted?

If it's agreed to, I'll rebase on 8.0.4xx.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs (hunk 2 above)
Comment (Forgind) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:151]: Note that this doesn't change based on an extraneous File.Exists check, so if you want a cleaner error message for the user for malformed install state files, or perhaps some kind of fallback (empty install state?), I can implement that, but it doesn't already exist in 400, so without customer need, I don't think that meets the bar for servicing.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs [hunk 4]
```diff
@@ -169,28 +169,24 @@ bool TryGetWorkloadSet(string workloadSetVersion, out WorkloadSet? workloadSet)
...
+                _manifestsFromInstallState = installState.Manifests is null ? null : WorkloadSet.FromDictionaryForJson(installState.Manifests!, _sdkVersionBand);
             }
 
-            if (_workloadSet == null && availableWorkloadSets.Any())
+            if (_workloadSet == null && installState.UseWorkloadSets == true && availableWorkloadSets.Any())
```
Comment (dsplaisted) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:189]: `UseWorkloadSets` is a nullable bool.  I think if it's null (IE not explicitly set), then we probably want the existing behavior where it will use workload sets.  That might also make the test changes unnecessary.

```suggestion
            if (_workloadSet == null && installState.UseWorkloadSets != false && availableWorkloadSets.Any())
```

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs [hunk 5]
```diff
@@ -169,28 +169,24 @@ bool TryGetWorkloadSet(string workloadSetVersion, out WorkloadSet? workloadSet)
...
+            var installState = InstallStateContents.FromPath(_installStateFilePath);
             if (_workloadSet is null)
             {
-                var installStateFilePath = Path.Combine(WorkloadInstallType.GetInstallStateFolder(_sdkVersionBand, _sdkRootPath), "default.json");
-                if (File.Exists(installStateFilePath))
+                if (!string.IsNullOrEmpty(installState.WorkloadVersion))
                 {
-                    var installState = InstallStateContents.FromPath(installStateFilePath);
-                    if (!string.IsNullOrEmpty(installState.WorkloadVersion))
+                    if (!availableWorkloadSets.TryGetValue(installState.WorkloadVersion!, out _workloadSet))
```
Comment (dsplaisted) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:178]: Was this change intentional?  Maybe it was a rebase/merge issue.  I think we want to call `TryGetWorkloadSet` here.

```suggestion
                    if (!TryGetWorkloadSet.TryGetValue(installState.WorkloadVersion!, out _workloadSet))
```

Comment (Forgind) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:178]: Ah, yes. Sorry; missed that when I was merging the branches. Will revert.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs (hunk 4 above)
Comment (dsplaisted) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:189]: Yes, I was also thinking it probably wouldn't affect the behavior in most cases.

I think the main argument for changing it is so that we don't have to remember to go back and change the code here when we change the default to use workload sets.  Another way to achieve that could be to add an `InstallStateContents.UseWorkloadSetsDefaultValue` constant, and wherever we read the `UseWorkloadSets` property, do something like `installState.UseWorkloadSets ?? InstallStateContents.UseWorkloadSetsDefaultValue`.

**Review by dsplaisted — APPROVED** [review]
It would be good to have some tests for this, be they functional and/or VM-based MSI tests.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs (hunk 4 above)
Comment (Forgind) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:189]: I support having a default value like that, but it's a bit late for 8.0.4xx, and hopefully we won't change the default after we change it in 9.0.1xx
